### PR TITLE
RFC: Suggest a "dot-aware" Entity::get() implementation.

### PR DESCRIPTION
**Pull request by beporter**
This PR demonstrates the ability to use `Entity::get('property.subproperty')` to reach deeply into an Entity. This works in the case where `$entity->property` is either an Entity itself or an array. It preserves all existing functionality and (in the unit tests are any indication) should be fully backwards compatible. It simply extends the interface to recognize and attempt to recursively support dotted-notation when the top-level retrieved property supports `::get()` or `Hash::get()` itself. Like all existing Entity "get" interfaces, this methods returns null in any failure case.

Arguments in favor:
- `Hash::get()` provides a dot notation interface for traversing deep into an array structure. This PR provides a similar interface to Entities.
- There is no (simple) way to  _programmatically_ use arrow notation (`$entity->property->subproperty`) to retrieve a "path" from nested objects, which limits the ability to use Entities dynamically.

If the discussion is positive, I will happily add tests and documentation.

Additional talking points:
- [x] Is `method_exists('get')` the best approach? Maybe `instanceof EntityInterface` would be better?
- [x] Adding tests of the **new** functionality may illuminate shortcomings or gotchas (but I'm not going to waste my time unless there is interest in this.)

### CODE REVIEW

**Comment by josegonzalez** [comment]
What happens if a developer provides a backend with a `json` field type (such as postgres)?

**Comment by beporter** [comment]
@josegonzalez Can you give me an example? Are you saying that using dotted properties names was officially supported in Entities already even though they wouldn't be accessible via `$entity->property.whatever`?

**Comment by josegonzalez** [comment]
I guess I'm hopeful this helps users implementing postgres json support - If someone wants to sponsor me doing this, I'd love to see it land in master! - so I was curious if that was the goal of this sort of project. The data would be represented as an array, so it seems like the answer is yes.

One other minor concern is that people put dots in their fieldnames. I have *no* idea if thats even valid sequel, but I wouldn't be surprised if someone has an app running around like that.

**Comment by beporter** [comment]
Yes, if they put dots in their fields names, they would have already had a bad time in Cake 3 though, since you can't access an Entity's properties directly by name. This would potentially conflict with that, but no worse so than before.

As for the json access, I haven't tested it, but I imagine that if you provide a mechanism for marshaling your json data into an array or a PHP object that implements its own `::get()` interface that itself can handle dotted-paths, then yes: the current implementation in this PR should be able to "dig into" those sub-properties. If the property value is just a json-encoded string though, then you'd get the same result as the EntityInterface guaranteed before this PR: `null`.

**Comment by phpnut** [comment]
@beporter +1 from me
@josegonzalez What do you means someone sponsor you?

**Comment by josegonzalez** [comment]
I was kidding :D

**Comment by beporter** [comment]
Truth be told, the (immediate) motivation behind this is because I wanted to write a unit test like this:

```php
$input = [
	// ...
];
$expected = Hash::merge($input, [
	'foo' => 'bar',
	'user' => [
		'role' => 'admin',
		'is_active' => true,
		'email_pending_verification' => 'foo@bar.com',
	],
];

$entity = $this->SubjectUnderTest->returnsAnEntity($input);

foreach (Hash::flatten($expected) as $path => $value) {
	$this->assertEquals($value, $entity->get($path));
}
```

File: src/Datasource/EntityTrait.php [hunk 1]
```diff
@@ -281,8 +299,10 @@ public function &get($property)
 
         if ($this->_methodExists($method)) {
             $result = $this->{$method}($value);
-            return $result;
+            return $recursor($result, $nested);
```
Comment (dereuromark) [inline src/Datasource/EntityTrait.php]: Cant you just remove the return statement? Looks like $recursor($result, $nested); will be called the next line anyway then.

Comment (beporter) [inline src/Datasource/EntityTrait.php]: That breaks the return-by-reference required by array-style reference. (Trust me, phpunit lit up like a :christmas_tree: when I tried to simplify further.)

Comment (dereuromark) [inline src/Datasource/EntityTrait.php]: OK :)

Comment (beporter) [inline src/Datasource/EntityTrait.php]: Also note that the final `return` operates on `$value` where the nested return above it operates on `$result`.

**Comment by beporter** [comment]
Is it safe to say there's at least enough interest that I should extend the test suite enough to start looking for problems?

**Comment by josegonzalez** [comment]
:+1:

**Comment by dereuromark** [comment]
Sure thing.

**Comment by beporter** [comment]
Okay, I'll get to work on that. Thanks all.

File: src/Datasource/EntityTrait.php [hunk 2]
```diff
@@ -281,8 +299,10 @@ public function &get($property)
...
         }
+
+        $value = $recursor($value, $nested);
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Is there a way we can avoid the additional function calls when the property access is a simple non-dotted path? While it might not seem like much all these extra function calls add up in big apps. I remember that part of the big speedups from 1.3 -> 2.0 was due to calling a few 1000 functions fewer per request.

**Comment by markstory** [comment]
One concern I have with this is it breaks the symmetry get() and set(). Would it also be possible to implement dotted paths for set() ?

File: src/Datasource/EntityTrait.php (hunk 2 above)
Comment (beporter) [inline src/Datasource/EntityTrait.php]: Probably yes. My first goal was to make the minimum modifications necessary to maintain complete backwards compatibility and have the test suite pass. The next step is to build the test suite out against this initial implementation to "cement" the expectations. Then as a final step it becomes safe to refactor the original code, using the new tests to ensure any changes still conform to the same expectations. The tests I'm adding are generally going well, though I won't have time to finish them tonight. That means probably tomorrow I expect to be able to submit an updated algorithm that short-circuits as quickly as possible in all cases.

Comment (markstory) [inline src/Datasource/EntityTrait.php]: Ok.

**Comment by beporter** [comment]
> One concern I have with this is it breaks the symmetry get() and set(). Would it also be possible to implement dotted paths for set() ?

I thought about it, but it's definitely more difficult to get "right." For one thing, `set()` can't just default to silently assigning `null` when it encounters a failure case like `get()` can, but in order for set to act on a dotted path, every dotted element would have to implement its own `::set()` interface and there's seriously no way to guarantee that. So what do we do when we can't set() all the way down the dotted path? Return false? Throw an exception? There is precedent for this in set() already of course, but it's not up to me to decide if adding more exception cases is a "good" change.

If that's the way you'd like to go I can take a shot at implementing it, but (without having tried it) right now I'm not exactly optimistic that it's possible to implement deep set() reliably. In addition, even `Hash` doesn't implement a set() (probably because it's hard?). That's a problem because I'm leveraging Hash::get() internally here to handle deep array access so a theoretically-expanded `Entity::set()` would need a counterpart `Hash::set()` to assist with deep array assignment. Seems like this gets ugly pretty quickly...

On the other hand, while I _personally_ would be willing to live with the mismatch between get() and set(), I could envision a lot of support requests for why `$entity->set('meaning.life', 42)`, "doesn't work," so whatever the consequences of that possibility are, I'm fine with it.

First things first though-- at least let me finish a "completed" `::get()` that addresses the RFC and the points raised so far. I suggest we revisit expanding the idea to set() afterwards.

**Comment by beporter** [comment]
In the meantime, here's an unusual side-effect I encountered while writing tests. This works (at least in PHP 5.6):

```php
echo $entity->{'property.sub.really.deep'};
// same as $entity->property->sub->really->deep;
```

Mind you, I'm not judging whether that's "good" or not...

**Comment by ADmad** [comment]
> Is method_exists('get') the best approach? Maybe instanceof EntityInterface would be better?

Checking `instanceof EntityInterface` would be better. That's what's done in other methods of the trait.

**Comment by ADmad** [comment]
@beporter What would be the benefit of using `$entity->get('prop.sub.deep')` v.s. `Hash::get($entity, 'prop.sub.deep')`  (besides a tad better usage syntax)?

**Comment by beporter** [comment]
@ADmad Well, Hash only recently gained the ability to be used with Entities (via the ArrayAccess interface, but you already know this, according to the commit history.)

I had original tried using Hash for this purpose in the 3.0.0 days, and "learned" that it didn't work with Entities. I never even thought to try again. The fact that Hash can do it now is great-- What that means is that the **only** advantage this PR could even offer is that "tad better usage syntax" you mentioned. But the good news is that I think we can probably farm out all of the heavy lifting directly to Hash instead of duplicating an implementation here.

**Comment by beporter** [comment]
...that is, if everyone thinks that having the dotted syntax available via Entity::get() is worth it.

In other words: The changes proposed in this PR don't provide any benefit over using Hash::get(), _other than_ the convenience of the syntax. If we don't care about that convenience, then we can just close this PR and tell people to use `Hash::get($entity, $path)` instead of the proposed `$entity->get($path)`.

**Comment by beporter** [comment]
The diff from the as-released 3.1.5 implementation would be something like this:

```diff
diff --git a/src/Datasource/EntityTrait.php b/src/Datasource/EntityTrait.php
index de42200..19506d8 100644
--- a/src/Datasource/EntityTrait.php
+++ b/src/Datasource/EntityTrait.php
@@ -272,6 +272,11 @@ trait EntityTrait
             throw new InvalidArgumentException('Cannot get an empty property');
         }
 
+        if (strpos($property, '.') > 0) {
+            $result = Hash::get($this, $property);
+            return $result;
+        }
+
         $value = null;
         $method = '_get' . Inflector::camelize($property);
```

That's a lot cleaner.

**Comment by ADmad** [comment]
@beporter Yeah I am guilty of adding that enhancement, hence my question as I wondered why all the complexity of your patch was needed :)

I am :+1: for having support for dotted syntax in `Entity::get()`. I don't think one extra `if` check as shown in your diff will add enough overhead to deny having this convenience.

**Comment by beporter** [comment]
@ADmad I'm just glad you mentioned it. It's a little scary that I might have not only duplicated the effort but complicated a pretty heavily used piece of the core entirely without good cause.

At least some of that has to be a communications thing though. I already mentioned that I didn't know Hash could do deep fetching from Entities, and my impression is that it's going to come as a surprise to some of the other core team people in this thread too. If you hadn't been monitoring, my first version might have gone in. That wouldn't have been the end of the world if it was caught later, but that kind of entropy (even when unintentional) rots a codebase over time otherwise.

Okay, so I'll fix up the PR to just leverage Hash and reduce my test cases to a more simple, "can we engage `Hash` like we expect?" since we don't need to re-test that functionality.

**Comment by ADmad** [comment]
> At least some of that has to be a communications thing though.

Seems I forgot to update the docs about the enhancement, my bad. I will rectify that soon unless you beat me to it :)

> my impression is that it's going to come as a surprise to some of the other core team people in this thread too

At least 2 devs who commented on this PR have commented on my enhancement PR for `Hash::get()` too, so probably they just forgot. Plus as you probably already know all team members get notifications for each PR and issue (unless they explicitly disabled them).

**Comment by beporter** [comment]
I've simplified the code as discussed and added test cases. The builds are currently running. A bonus of having a dot-aware `Entity::get()` is that we also get a dot-aware `Entity::has()`.

I also added a quick test to cover the one remaining untested line in the file from `::getOriginal()`.

**Comment by dereuromark** [comment]
Thats what I am talking about - now it looks really neat and awesome :)

**Comment by beporter** [comment]
@dereuromark: Credit for that goes to @ADmad. I'm just using the work he did in `Hash` and didn't even know about. :sweat_smile:

**Comment by beporter** [comment]
Is there anything else we're waiting on here?

**Comment by markstory** [comment]
In response to [this comment](https://github.com/cakephp/cakephp/pull/7847#issuecomment-164628197)

> In addition, even Hash doesn't implement a set() (probably because it's hard?).

There is a `Hash::insert()` method that effectively does 'set'.

> On the other hand, while I personally would be willing to live with the mismatch between get() and set(), I could envision a lot of support requests for why $entity->set('meaning.life', 42), "doesn't work," so whatever the consequences of that possibility are, I'm fine with it.

One issue with the above example will be that `meaning.life` key will never be readable with this change in place.

I do agree that universal set() is more difficult than get(). I think that we can cover all the cases without too many exceptions though by following Hash::insert(). 

* Scalar values are cast to an array and then modified.
* Arrays are modified fine.
* Objects that implement ArrayAccess are modified in place.
* Objects without ArrayAccess would cause an exception. I don't think we want to go down the path of trying to guess how values should be set into arbitrary objects.

**Comment by beporter** [comment]
I'm not super-familiar with `Hash::insert()`, but reading the code it seems like it will do what we want (for arrays). It seems like the implication of your suggestion is to attempt to piggy back on that behavior for `Entity::set()` as well, is that correct?

If so, the typehint for Hash::insert() might need to change to support `ArrayAccess` instead of only `array`.

Beyond that, Hash::insert() doesn't (currently) do what you'd expect when it comes to Entities. Playing around on the Console, I tried this:

```php
$e = new \Cake\ORM\Entity([
	'sub' => new \Cake\ORM\Entity([
		'deep' => 1,
		'second' => 2,
	])
]);
$e = Hash::insert($e, 'sub.third', 3);
debug($e);
```

Which produces this (trimmed) output:

```
/src/Shell/workingShell.php (line 34)
########## DEBUG ##########
object(Cake\ORM\Entity) {
	'sub' => [
		'third' => (int) 3
	],
}
###########################
```

So without having played around at all yet, Hash::insert() is replacing the entire `$e->sub` property with an array that looks like this: `['third' => 3]`. More experimentation is warranted here.

> One issue with the above example will be that meaning.life key will never be readable with this change in place.

As a counter-point: That key would only be readable in existing versions of Cake if it had been created as a top-level property that included the dot, which I suspect is pretty unlikely or we wouldn't be considering this change to begin with. With the enhancements in this PR, it would become a `life` sub-property of the `meaning` key in the top-level Entity, that's true. If you're speaking about having used `set()` to assign the value as a "top-level property with a dot in the name" and then not being able to pull it back out via `get()`, then yes; it might be prudent for this PR to include a thrown exception when calling set() with a dotted path (or as you're suggestion, implementing dotted notation for set() as well.)

Alternatively, we can guard the backwards-compatible case. We could specifically check for a top-level key by changing the new if statement thusly:

```diff
-if (strpos($property, '.') > 0) {
+if (strpos($property, '.') > 0 && !isset($this->_properties[$property])) {
         $result = Hash::get($this, $property);
         return $result;
 }
```

Basically: _"Only engage Hash::get() if we're sure there isn't a top-level key of the same name."_

A dot isn't allowed in a custom `_getWhatever()` method name thanks to PHP's language rules already (and [how the Inflector::camelize() treats dotted names](http://inflector.cakephp.org/first.second)), so there's no worry of breaking compatibility there.

---

So it looks like where that leaves us is that I need to play around with trying to inject `Hash::insert()` support into `Entity::set()`, correct?

**Comment by beporter** [comment]
Other random things to think about and work through (mostly for my own reference):

* `Hash::insert()` returns the modified array. `Entity::set()` changes properties on `$this` "in-place". We can't just mirror what we're doing for `::get()`:
    ```php
    if (strpos($p, '.') > 0) {
        $this = Hash::insert($this, $p); // nope
    }
    ```

* `Entity::set()` also stores `::$original` values. If we are recursing into nested Entities, we have to make sure this engages at the deepest Entity (and up the chain?).

* `::dirty()` probably needs to be set all the way down the chain as well (but that doesn't necessarily mean that it needs to support dotted-notation directly itself.)

* In order to ensure compliance with `::$_accessible` fields, we'll potentially have to make `Entity::accessible()` dot-aware.

* This isn't an issue from a code perspective, but an implication of allowing `Entity::set()` to handle dotted paths would be multiple assignment like this:

    ```php
    $entity->set([
        'meaning.life' => 42,
        'article.comment.author' => 'Mark',
    ]);
    ```

    ```
    object(Cake\ORM\Entity) {
        'meaning' => object(Cake\ORM\Entity) {
            'life' => 42,
        },
        'article' => object(Cake\ORM\Entity) {
            'comment' => object(Cake\ORM\Entity) {
                'author' => 'Mark',
            },
        },
    }
    ```
* Assigning to a "deep" property that doesn't exist as a complete path must continue to "auto-create" array sub-keys in the deepest-level entity as `Hash::insert()` already does.

**Comment by markstory** [comment]
>  If you're speaking about having used set() to assign the value as a "top-level property with a dot in the name" and then not being able to pull it back out via get()

Yes that is exactly my concern.

I totally forgot about original and dirty state when suggesting `Hash::insert()`. It might not be worth the effort to try and use `Hash::insert()` with that information.

In my mind, setting deep properties would create new arrays. I don't think we reasonably make any guesses at concrete classes, and an array is not 'wrong'.

**Comment by beporter** [comment]
There are actually some nitty-gritty points/questions that come up while trying to implement this. Take this **unfinished** implementation:

```diff
diff --git a/src/Datasource/EntityTrait.php b/src/Datasource/EntityTrait.php
index a038864..6ca965e 100644
--- a/src/Datasource/EntityTrait.php
+++ b/src/Datasource/EntityTrait.php
@@ -232,12 +232,29 @@ trait EntityTrait
     public function set($property, $value = null, array $options = [])
     {
         $isString = is_string($property);
         if ($isString && $property !== '') {
             $guard = false;
             $property = [$property => $value];
         } else {
             $guard = true;
             $options = (array)$value;
         }

         if (!is_array($property)) {
             throw new InvalidArgumentException('Cannot set an empty property');
         }
         $options += ['setter' => true, 'guard' => $guard];
 
         foreach ($property as $p => $value) {
+            $nested = false;
+            if (strpos($p, '.') > 0 && !array_key_exists($p, $this->_properties)) {
+                list($p, $nested) = explode('.', $p, 2);
+            }
+
             if ($options['guard'] === true && !$this->accessible($p)) {
                 continue;
             }
 
             $this->dirty($p, true);
 
+            if ($nested) {
+                if ($this->_properties[$p] instanceof EntityInterface) {
+                    $this->_properties[$p] = $this->_properties[$p]->set($nested, $value, $options);
+                } elseif ($this->_properties[$p] instanceof ArrayAccess) {
+                    $this->_properties[$p] = Hash::insert((array)$this->_properties[$p], $nested, $value);
+                } else {
+                    throw new InvalidArgumentException('Cannot set dotted property');
+                }
+
+                continue;
+            }
+
             if (!array_key_exists($p, $this->_original) &&
                 array_key_exists($p, $this->_properties) &&
                 $this->_properties[$p] !== $value
             ) {
                 $this->_original[$p] = $this->_properties[$p];
             }

             if (!$options['setter']) {
                 $this->_properties[$p] = $value;
                 continue;
             }

             $setter = '_set' . Inflector::camelize($p);
             if ($this->_methodExists($setter)) {
                 $value = $this->{$setter}($value);
             }
             $this->_properties[$p] = $value;
         }

         return $this;
     }
```

* As discussed before, any "pre-existing" properties with dotted names are honored, partially preserving backwards compatibility (how you'd end up with an Entity containing a property name with dots in it to begin with is now the difficult part.) 

* The `accessible()` state of **only the first dotted portion** is now checked (agin keeping in mind this only happens when there isn't an existing top-level property name matching the entire dotted path):

    ```php
    $entity->accessible('top', false);
    $entity->set('top.second.third', 'new value'); // blocked at "top"
    ```

* `dirty()` is set as far down the dotted path as we're able to traverse. This makes sense to me in the context of what needs to be saved back to a datasource, but correct me if I'm wrong.

* We recognize our own `EntityInterface` and propagate calls to `->set()` using the `$nested` path. This  is what propagates dirty() and honors accessible(), so we don't have to make those methods dot-aware directly.

* If we don't have an EntityInterface to work with, we check for `ArrayAccess` instead and hand-off to `Hash::insert()` for that case. I think this preserves "expected behavior" as much as possible and fulfills Mark's outline pretty well by re-using existing behavior.

    The `(array)` cast is currently necessary since `Hash::insert()` typehint's to `array` and not `ArrayAccess`, but we can change that too if necessary.

* As Mark suggested, we explicitly throw an `InvalidArgumentException` when trying to set to a dotted path where that path can't be supported. This would potentially break existing applications with code like this: `$entity->set('really.a.top.level.property', 'someVal');` but _only where `$entity->_properties['really.a.top.level.property']` didn't **already** exist in the Entity._ 

    The alternative is to remove the `elseif(ArrayAccess)` and the `else InvalidArgument` and just always cast to `(array)` whenever we don't have an `EntityInterface`. This could cause actual (and unexpected) changes to the data type being stored though.

* The existing `EntityTest.php` class passes with the above changes, but I haven't added any new tests to exercise the above additions yet.

**Comment by markstory** [comment]
I don't know what the expected behavior of setter methods should be when a nested property is set to a non-entity property. Should we trigger setter methods for just the root property, and pass the nested structure in?

**Comment by beporter** [comment]
Actually, the opposite kind of makes sense to me. If you think of the dotted path as "just traversal," then **only** engaging the setter at the bottom of the chain makes sense. "I want to dig into this nested path, and set this property on this sub-sub-sub-entity (using any available setter).

**Comment by beporter** [comment]
For example, I would want it to behave the same as if I named the properties explicitly:

```php
$entity->set('subEnt.deeperEnt.reallyDeep.superDeep', 'my value');
// should be equivalent to:
$entity->subEnt->deeperEnt->reallyDeep->set('superDeep', 'my value');
```

None of the intermediate setters would be called in that case.

**Comment by markstory** [comment]
@beporter Good point. I was thinking about the operation from the root entity's perspective which is not as good as your point of view.

**Comment by beporter** [comment]
Status update: I have some time in on the modifications to `::set()` but work is currently taking priority. Shooting for more progress this weekend.

**Comment by beporter** [comment]
Just a summery note on this since there was some recent activity in #8619: The `Entity::get()` interface is dot-ready, but the consensus was to try to also add support for dot-ready `Entity::set()` at the same time, which I have not had time to complete. (I have a few experimental commits on that done locally that I have not finished or pushed yet.) My goal is to have it done by CakeFest so nobody has to yell at me in person. :sweat_smile:

**Comment by ADmad** [comment]
> My goal is to have it done by CakeFest so nobody has to yell at me in person.

@beporter So are you getting near to completing changes for `Hash::set()` or should we start getting the yelling party ready? :stuck_out_tongue: 

Seriously though if changes to `Hash::get()` are complete you can rebase your branch and we can merged this after a final review. Changes to `Hash::set()` can be done later whenever you (or some else) gets the time.

**Comment by beporter** [comment]
> So are you getting near to completing changes for Hash::set() or should we start getting the yelling party ready?

Hey, I still have a week left! 😝 

> Seriously though if changes to Hash::get() are complete you can rebase your branch and we can merged this after a final review. Changes to Hash::set() can be done later whenever you (or some else) gets the time.

Yep I like that! I will make that my first goal, with "completing the set() interface" as the follow up.

**Comment by beporter** [comment]
Build is running, but there are some final thing I want to bring up. (There are related comments I've left in the code for the time being to help facilitate this conversation):

* In the `$entity->set('nested.path.here', 1);` case where `$entity->nested` does not yet exist, it turns out that `Hash::insert()` is not the best option for [autovivication](https://en.wikipedia.org/wiki/Autovivification) of sub-records. Arrays are "lossy" in the sense that they can't track original values, field errors, etc. This creates a impedance mismatch between get() and getOriginal() that I couldn't resolve easily, so I used Entities instead.

* As discussed earlier in this thread, `$entity->get('nested.path.here');` **will not engage getters** until it reaches the `->here` property at the end. The rest of this operation is simply node traversal. This behavior makes `$e->get('nested.path') === $e->nested->path`.

* An exception is thrown when attempting to set to a nested path "past" a scalar value. Consider:

    ```php
    $e->strVal = 'I am a string';
    $e->set('strVal.nested', 'Strings are not objects or arrays!); // Throws InvalidArgumentException
    ```

    This seems most reasonable to me, since there are other examples of throwing exceptions in the trait already in the Trait.

* I see the (slim) potential for reducing code duplication. There are snippets in `::get()`, `::set()` and `::getOriginal()` now that are quite similar for dealing with the nested paths.

* The book definitely needs a PR to explain this change, in particular the case when trying to assign deeply into a scalar value noted above, as well as a note about not using dots in your property names (I still contend that this aspect of the PR alone makes it more suited for 3.3.0 at this point.)

* I completed test coverage for `EntityTrait`, adding cases for the `::invalid()` and `getOriginal()` methods.

**Comment by beporter** [comment]
Quick Q: I'm working on adding to the book to explain this new feature. Which branch of the docs should I target? Possible candidates seem to be:

* `master`
* `3.0`
* `3.3`

File: src/Datasource/EntityTrait.php [hunk 3]
```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
...
             }
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
```
Comment (lorenzo) [inline src/Datasource/EntityTrait.php]: Should we just use an `ArrayObject` instead?

Comment (lorenzo) [inline src/Datasource/EntityTrait.php]: What I'm not comfortable with is coupling it to a class that is not in this sub-repo.

File: src/Datasource/EntityTrait.php [hunk 4]
```diff
@@ -283,6 +320,23 @@ public function &get($property)
...
+
+        if ($nested) {
+            if ($this->_properties[$property] instanceof EntityInterface) {
+                $result = $this->_properties[$property]->get($nested);
+                return $result;
+            } elseif (is_array($this->_properties[$property])
+                || ($this->_properties[$property] instanceof \ArrayAccess)
+            ) {
+                $value = Hash::get($this->_properties[$property], $nested);
+                return $value; //@TODO: Or remove this line and let getters engage below?
```
Comment (lorenzo) [inline src/Datasource/EntityTrait.php]: No, this is fine... it should not continue further down

**Comment by lorenzo** [comment]
Is there a test case for something like this?

```php
$entity = new Entity(['my_prop' => 5]);
$entity->set('my_prop.thing', 'something else');
```

What will happen in that case? I'm more inclined to say that we need to throw an exception

File: src/Datasource/EntityTrait.php (hunk 3 above)
Comment (beporter) [inline src/Datasource/EntityTrait.php]: That's a good point, but the issue is the impedance mismatch between ::getOriginal and ::set(). When you set a dotted path that must be created, if you create the sub-objects as arrays or ArrayObjects, you can't retain `::$_original` data for those new sub items. It' be _possible_ to discard this information, but it will actually make the implementation in set() and getOriginal() more complex. I'm fine with whatever, but I did actually start with ArrayObjects and realized that path was going to be more complicated.

File: src/Datasource/EntityTrait.php (hunk 4 above)
Comment (beporter) [inline src/Datasource/EntityTrait.php]: The argument for doing that is: "the last applicable getter should fire."

Consider:

```php
$entity = new Entity([
    'a' => [
        'b' => 'value'
    ],
]);
echo $entity->a['b']; // Engages `_getA()`, if it exists.
echo $entity->get('a.b'); // Should probably still call `_getA()` before using `Hash::get('b')`??
```

**Comment by beporter** [comment]
@lorenzo There is in fact a test case for that, and it does in fact throw an Exception since you can't descend into a scalar value. The current exception message would read like this for your example:

```
Cannot set dotted property when a traversal segment contains a scalar value: my_prop
```

See `EntityTest::testSetDottedScalarPath()`.

File: src/Datasource/EntityTrait.php (hunk 3 above)
Comment (beporter) [inline src/Datasource/EntityTrait.php]: Here's an illustration:

```php
$e = new Entity(['a' => null]);

// If we use ArrayObjects (or arrays):
$e->set('a.b.c', true); // Internally: ::_properties['a'] = ['b' => ['c' => true]];
$e->getOriginal('a.b.c'); // Either throws an Exception or gives you all of `a`.

// If we use Entities:
$e->set('a.b.c', true); // Internally: ::_properties['a'] = new Entity(['b' => new Entity(['c' => true])]);
$e->getOriginal('a.b.c'); // Works.
```

What concerns me is being able to call `::set('a.b.c')` and have it work, but **not** being able to call `::getOriginal('a.b.c')` in all of the same cases that set() does. _Sometimes_ it will work (if a, b and c are all Entities) and sometimes it won't. But of course it won't work either way for existing properties that are already arrays, so maybe I'm being paranoid?

Comment (beporter) [inline src/Datasource/EntityTrait.php]: I'm starting to think that throwing an Exception in `::getOriginal()` when the requested path is dotted but ::_original data is not available all the way down the chain is the best course. Basically saying, "don't try that, because we can't answer your question in this case." There may be side-effects that I'm not anticipating with that yet, but lemme poke at the tests here a bit.

File: src/Datasource/EntityTrait.php (hunk 4 above)
Comment (lorenzo) [inline src/Datasource/EntityTrait.php]: Hmm, true... It should probably use the getter for `b` after getting the value for `a` by using its getter as well.

Comment (beporter) [inline src/Datasource/EntityTrait.php]: But remember that `b` isn't a direct property on `$entity` in this case. Even if we had an `$entity::_getB()` method, it would apply to `$entity->b` and not to `$entity->a['b']`. In this case, `$entity::_getA()` is the last possible getter method involved since we switch over to `Hash::get()` in order to retrieve `b` from `a`. If `$entity->a` was _also_ an Entity, then the code already descends deeper and returns `$entity->a->_getB()` instead. Although...

I need to refactor and experiment a bit. I'm rethinking an [earlier conversation with @markstory about engaging setters recursively](https://github.com/cakephp/cakephp/pull/7847#issuecomment-168219991). In the case of set(), we should use get() all the way down to the last Entity or array possible before calling ::set() or Hash::set(). But with ::get(), we should be invoking every possible getter along the chain:

```php
// These two are already equivalent (assuming the getter methods are present):
$entity->a->b->c;
$entity->_getA()->_getB()->_getC();

// We want this to **also** be equivalent:
$entity->get('a.b.c');
```

Comment (lorenzo) [inline src/Datasource/EntityTrait.php]: Agreed, the third case should be equivalent to the previous two

File: src/Datasource/EntityTrait.php (hunk 3 above)
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Couldn't `getOriginal()` bail on objects that don't implement `EntityInterface`? We must have similar code for handling arrays.

File: src/Datasource/EntityTrait.php [hunk 5]
```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
...
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
+                    //$value = Hash::insert([], $nested, $value); //@TODO: Hash::insert() is the other option to generate arrays down the chain instead of Entities, but this has drawbacks like not preserving the ability to ::getOriginal() down the chain. The most fidelity is preserved by using Entities.
```
Comment (markstory) [inline src/Datasource/EntityTrait.php]: More important than `getOriginal()` is the dirty tracking. Without dirty tracking we won't be able to properly persist entities.

File: src/Datasource/EntityTrait.php (hunk 4 above)
Comment (markstory) [inline src/Datasource/EntityTrait.php]: Engaging getters/setters seems like it would be least surprising. It also has the nice effect of making

```php
$article->get('author.group.name');
```

into 'syntactic sugar' for

```php
$article->author->group->name;
```

File: tests/TestCase/ORM/EntityTest.php [hunk 6]
```diff
@@ -46,6 +46,31 @@ public function testSetOneParamNoSetters()
         $this->assertSame(1, $entity->id);
         $this->assertEquals(1, $entity->getOriginal('id'));
         $this->assertEquals('bar', $entity->getOriginal('foo'));
+
+        $entity->set('level1.level2.level3', 'three');
+        $this->assertSame('three', $entity->level1->level2->level3);
+        $this->assertEquals('three', $entity->getOriginal('level1.level2.level3'));
+        $entity->set('level1.level2.level3', '333');
+        $this->assertEquals('three', $entity->getOriginal('level1.level2.level3'));
```
Comment (markstory) [inline tests/TestCase/ORM/EntityTest.php]: Assertions for the dirty state would be good to have too.

File: src/Datasource/EntityTrait.php (hunk 3 above)
Comment (beporter) [inline src/Datasource/EntityTrait.php]: Agreed, that's what it's looking like. It'll _have_ to throw a new exception since there's no other logical outcome if you ask for a nested path that descends into something that doesn't implement `getOriginal()` (EntityInterface) itself. That will also make it easier to switch back to ArrayObjects instead of Entities I think. Stand by please...

File: tests/TestCase/ORM/EntityTest.php (hunk 6 above)
Comment (beporter) [inline tests/TestCase/ORM/EntityTest.php]: It's worth noting that like "original" values, dirty state can only descend down to the last available entity.

```php
$e = new Entity(
    ['a' => ['b' => 'fun']], 
    ['markClean' => true]
);
var_dump($e->dirty('a')); // false

$e->set('a.b', 'times'); // Make an array element dirty.
var_dump($e->dirty('a')); // Definitely true still.

// True (because we stop at 'a' being dirty)? 
// Or Exception (because we can't descend into 'b' to _check_ for dirty)??
var_dump($e->dirty('a.b')); // ???
```

Comment (markstory) [inline tests/TestCase/ORM/EntityTest.php]: I think we could would want to descend if the nested objects are entities, or take 'last known information' for non-entity children.

Comment (beporter) [inline tests/TestCase/ORM/EntityTest.php]: Totally agree, but the question is _how_ to stop when we run out of Entities (but could theoretically descend into an array that lacks the ability to track dirty state.) 

Is that an Exception or does that short-circuit at the farthest Entity down the chain and return true?

Comment (beporter) [inline tests/TestCase/ORM/EntityTest.php]: ...Or are you saying we should start storing more fine grained dirty state for **elements** inside of dirty properties?

Comment (markstory) [inline tests/TestCase/ORM/EntityTest.php]: > Totally agree, but the question is how to stop when we run out of Entities (but could theoretically descend into an array that lacks the ability to track dirty state.) 

In the EntityContext code I did [some sniffing](https://github.com/cakephp/cakephp/blob/master/src/View/Form/EntityContext.php#L268-L338) and poking to find the leaf entity object.

**Comment by beporter** [comment]
I'm gonna close this. I don't expect to be able to work around the issue with the `set()` side of the equation in the near future. `get()` was pretty trivial to add, but if we need to maintain parity there are just too many edge cases. I'll leave my branch around in case anyone want to try themselves.
